### DISCUSSION TITLE
detrending or filtering topography

### DISCUSSION THREAD

**Discussion by nombac** [ContactEngineering/SurfaceTopography#397] [Q&A]
Up until now, I've only been using SurfaceTopography for data loading, as shown in the code below. Now, I'd like to apply detrending and filtering to the loaded topography. I found the descriptions of the SurfaceTopography.Uniform.Detrending and SurfaceTopography.Uniform.Filtering modules in the manual, but I’m not sure how to use them  since I'm not very experienced with Python. Could someone explain how to implement these in the example code below? Thank you.

```
from SurfaceTopography import read_topography
import matplotlib.pyplot as plt

topo = read_topography('test.datx')
h = topo.heights()

plt.imshow(h)
plt.show()
```

**Comment by pastewka** [comment]
Hi @nombac - the detrending classes register themselves as method to the topography class. The idea is that you can chain together filters. For example, to do detrending to the following:

```python
topo = read_topography('test.datx')
detrended_topo = topo.detrend()
```

The `detrend` method return a new topography objects. (It's actually a pipeline that applies the detrending to the original topography, but that doesn't really matter for you.)

You can chain things together, for example:

```python
topo = read_topography('test.datx')
filtered_topo = topo.fill_undefined_data().detrend()
```

**Comment by pastewka** [comment]
You can find a list of pipeline functions here:

https://contactengineering.github.io/SurfaceTopography/usage.html

I've just updated the list to be more complete.

**Comment by nombac** [comment]
Thank you very much for the guidance!

I was considering applying LongCutTopography (https://contactengineering.github.io/SurfaceTopography/source/SurfaceTopography.Uniform.html#SurfaceTopography.Uniform.Filtering.LongCutTopography).

However, when I tried the following code:
```
filtered_topo = topo.longcut(cutoff_wavelength=1)
```
I encountered the following error:
```
    raise ValueError("only implemented for periodic topographies")
ValueError: only implemented for periodic topographies
```
As the error message suggests, is this an issue with the nature of the topography itself, rather than my usage of the filter?
If that’s the case, do you have any plans to extend the filter to handle non-periodic topographies as well?

**Reply by pastewka** [reply]
This is a Fourier filter which can't really be applied to nonperiodic data. One could window the data, apply the tilter, and then remove the window again by I've never tried this and it may garble the data.

What are you trying to do, i.e. why do you want to apply the filter? Maybe a Gaussian filter is what you want? (We don't have this implemented but it something that would be useful anyway.)

**Comment by nombac** [comment]
I’ve been using Zygo Mx software to filter topography data, specifically with the following settings:

- Filter Type: Spline
- Filter: High-pass
- Type: Gaussian Spline Auto

as shown below. I'm not very familiar with the terminology, so I wanted to ask if this is the Gaussian filter you mentioned. Also, do you have any insights on what "Spline" refers to in this context?

In any case, this filtering method works really well for revealing fine structures in my data. It would be fantastic if something similar could be implemented in SurfaceTopography!

<img width="888" alt="image" src="https://github.com/user-attachments/assets/58be2801-6020-4c4e-8a8b-9838a664130d">
